src/hyperdeck/response-handler.js

```diff
@@ -12,73 +12,78 @@ var ASYNC_RESPONSE_REGEX = /^(5[0-9]{2} )/;
-    // so remove it.
-    // i.e. "1\r\n2\r\n" => ["1", "2", ""]
-    lines.pop();
-    if (lines.length === 1) {
-      var asyncResponse = ASYNC_RESPONSE_REGEX.exec(lines[0]);
-      if (!asyncResponse) {
-        if (nextResponseIsMultiline) {
-          logger.debug('Assuming buffer is not complete because nextResponseIsMultiline flag is set.');
-          nextResponseIsMultiline = false;
-          return false;
-        }
-      }
-      // is it a single line response?
-      return lines[0].indexOf(":") !== lines[0].length-1;
-    } 
-    // multi line response so waiting for a blank line to signify end
-    return lines[lines.length-1] === "";
+  var buffer = [];
+
+  function isRespComplete() {
+    let complete = false;
+
+    if (buffer.length === 1) {
+      // a single line response
+      // single line response first line should never ends with ":" and
+      // 1XX and 200 is always single line response
+      complete = buffer[0].search(/^1\d\d\s|200\sok$/) === 0 && buffer[0].endsWith(':') === false;
+    } else if (buffer.length >= 3) {
+      // multi line response, so waiting for a blank line to signify end
+      complete = buffer[buffer.length - 1] === '';
+    }
+
+    return complete;
   }
-  
+
   function onData(rawData) {
-    logger.debug('Got data on socket.', rawData);
-    rawData.split("\r\n").forEach(function(line) {
-      if (buffer === '' && line.trim() === '') {
-        // handle empty lines before the data
-        // see https://github.com/LA1TV/Hyperdeck-JS-Lib/issues/44
-        return;
-      }
-      buffer += (line + "\r\n");
-      if (!isBufferComplete()) {
-        return;
+    logger.debug('Got data on socket.\n', rawData);
+
+    const resArray = rawData.split('\r\n');
+    // every response(even one line) must always end with \r\n, if not its not a valid response
+    if (resArray.length === 0 || resArray[resArray.length - 1] != '') {
+      throw "Unknown response type.";
+    }
+
+    resArray.pop();
+    resArray.forEach(function (line) {
+      // push to buffer till response is read completly
+      if (buffer.length > 0 || (buffer.length === 0 && line)) {
+        buffer.push(line);
       }
-      logger.debug('Got complete data.', buffer);
-      var data = Parser.parse(buffer);
-      // reset buffer
-      buffer = "";
-
-      switch (data.type) {
-        case "synchronousFailure":
-        case "synchronousSuccess":
-          var response = {
-            success: data.type === "synchronousSuccess",
-            data: data.data
-          };
-          notifier.emit("synchronousResponse", response);
-          break;
-        case "asynchronous":
-          notifier.emit("asynchronousResponse", data.data);
-          break;
-        default:
-          throw "Unknown response type.";
+
+      if (isRespComplete()) {
+        // remove the command seperator or extra last line
+        if (buffer[buffer.length - 1] === '') {
```

Comment:
I think it would be better to just do `if (buffer.length > 2)`. This is just handling removing the empty line after a multiline response.

Reply:
👍

Reply:
just to make sure you are not popping any data if its empty then pop.

Reply:
But isRespComplete() guatentees the last line will be empty if there are multiple lines. Also I meant > 1 not 2

Reply:
Yeah, still don't you think [ if (buffer[buffer.length - 1] === '') ] explains better than [ (buffer.length > 2) ] what you going to pop. ?

Reply:
Well `if (buffer[buffer.length - 1] === '')` implies that the last line might not be empty, which we know is always the case for multiline responses, so there's no reason to check.

Also just checking `buffer.length` will be marginally faster than reading `buffer.length`, finding that item, and then comparing it with `''`.

Reply:
of course, buffer.length will be faster but popping is a very important step and makes it double safe.

Reply:
I agree we need to pop(). I just think we only need to check the length is > 1 not that it is '' :)

Reply:
i meant we must never pop() any valid data, which will never be the case. but just for double safety.

Reply:
But we know that there is no data there. If there is data there then we have a bug :)

I think it's better to find out there is a bug, than be silently fixing it.

If it was more complex I'd do something like
```javascript
if (buffer.length > 1) {
  // multiline response, remove empty line
  var lastLine = buffer.pop();
  if (lastLine !== '') {
    throw new Error("Last line in multiline response wasn't empty?!");
  }
}
```

If we check that the line is empty then this implies it might not be, and this might complicate things if more work needs to be done later when someone else is looking at the code